Add signature parameter to inspect_types

jrbourbeau opened the pull request:
This PR adds a `signature=` parameter to the dispatcher `inspect_types()` method.

Fixes #4127

jrbourbeau commented:
Currently, the value of `signature` is ignored when `pretty` is `True`. One potential way to support this is to also add a `signature` parameter to the `Annotate` class constructor in `numba/pretty_annotate.py`. Does this seem like a reasonable way forward?

stuartarchibald commented:
Thanks for the patch @jrbourbeau, much appreciated.

> Currently, the value of signature is ignored when pretty is True. One potential way to support this is to also add a signature parameter to the Annotate class constructor in numba/pretty_annotate.py. Does this seem like a reasonable way forward?

Just had a quick look at the code in question. What you propose seems like a reasonable approach, if you start and it turns into a rabbit hole please shout :) Thanks.

jrbourbeau commented:
I added a `signature` parameter to `Annotate.__init__` and now `inspect_types(signature=...)` works for both `pretty=True` and `pretty=False`. 

One unexpected thing I ran into is that the dispatcher `get_annotation_info()` method returns information from the class level `func_data` attribute

https://github.com/numba/numba/blob/9774398f3d661600f1f76d7eb93dbfa73f6df0f9/numba/annotations/type_annotations.py#L56

Using this class level attribute means that calling `get_annotation_info` with a specified `signature` parameter returns info on all signatures for all compiled functions, not just for the specified signature. 

<details>
<summary>Example:</summary>

```python
from numba import njit
from numba.annotations.type_annotations import TypeAnnotation

@njit
def foo(x):
    return x + 1

foo(10)
foo(10.4)

print('\n First signature')
print(foo.get_annotation_info(signature=foo.signatures[0]))
print('\n Second signature')
print(foo.get_annotation_info(signature=foo.signatures[1]))
```
</details>

<details>
<summary>Output (on master):</summary>

```
 First signature
OrderedDict([(('_examples/typeannotation-global.py:6', '(int64,) -> int64'), {'num_lifted_loops': 0, 'filename': '_examples/typeannotation-global.py', 'funcname': 'foo', 'python_lines': [(5, '@njit'), (6, 'def foo(x):'), (7, 'return x + 1')], 'python_indent': {5: 0, 6: 0, 7: 4}, 'python_tags': {5: '', 6: '', 7: ''}, 'ir_lines': {5: [('label 0', '')], 6: [], 7: [('x = arg(0, name=x)  :: int64', ''), ('$const0.2 = const(int, 1)  :: Literal[int](1)', ''), ('$0.3 = x + $const0.2  :: int64', ''), ('del x', ''), ('del $const0.2', ''), ('$0.4 = cast(value=$0.3)  :: int64', ''), ('del $0.3', ''), ('return $0.4', '')]}, 'ir_indent': {5: [0], 6: [], 7: [2, 2, 2, 2, 2, 2, 2, 2]}})])

 Second signature
OrderedDict([(('_examples/typeannotation-global.py:6', '(int64,) -> int64'), {'num_lifted_loops': 0, 'filename': '_examples/typeannotation-global.py', 'funcname': 'foo', 'python_lines': [(5, '@njit'), (6, 'def foo(x):'), (7, 'return x + 1')], 'python_indent': {5: 0, 6: 0, 7: 4}, 'python_tags': {5: '', 6: '', 7: ''}, 'ir_lines': {5: [('label 0', '')], 6: [], 7: [('x = arg(0, name=x)  :: int64', ''), ('$const0.2 = const(int, 1)  :: Literal[int](1)', ''), ('$0.3 = x + $const0.2  :: int64', ''), ('del x', ''), ('del $const0.2', ''), ('$0.4 = cast(value=$0.3)  :: int64', ''), ('del $0.3', ''), ('return $0.4', '')]}, 'ir_indent': {5: [0], 6: [], 7: [2, 2, 2, 2, 2, 2, 2, 2]}}), (('_examples/typeannotation-global.py:6', '(float64,) -> float64'), {'num_lifted_loops': 0, 'filename': '_examples/typeannotation-global.py', 'funcname': 'foo', 'python_lines': [(5, '@njit'), (6, 'def foo(x):'), (7, 'return x + 1')], 'python_indent': {5: 0, 6: 0, 7: 4}, 'python_tags': {5: '', 6: '', 7: ''}, 'ir_lines': {5: [('label 0', '')], 6: [], 7: [('x = arg(0, name=x)  :: float64', ''), ('$const0.2 = const(int, 1)  :: Literal[int](1)', ''), ('$0.3 = x + $const0.2  :: float64', ''), ('del x', ''), ('del $const0.2', ''), ('$0.4 = cast(value=$0.3)  :: float64', ''), ('del $0.3', ''), ('return $0.4', '')]}, 'ir_indent': {5: [0], 6: [], 7: [2, 2, 2, 2, 2, 2, 2, 2]}})])
```
</details>

This behavior seems (naively) unexpected to me and made it difficult to support `inspect_types(signature=...)` when `pretty=True`. So I've modified `get_annotation_info` to only return information for the relevant signature when the `signature` parameter is specified. For example, running the above example with the changes in this PR gives

<details>
<summary>Output (on this branch):</summary>

```
 First signature
OrderedDict([(('_examples/typeannotation-global.py:6', '(int64,) -> int64'), {'num_lifted_loops': 0, 'filename': '_examples/typeannotation-global.py', 'funcname': 'foo', 'python_lines': [(5, '@njit'), (6, 'def foo(x):'), (7, 'return x + 1')], 'python_indent': {5: 0, 6: 0, 7: 4}, 'python_tags': {5: '', 6: '', 7: ''}, 'ir_lines': {5: [('label 0', '')], 6: [], 7: [('x = arg(0, name=x)  :: int64', ''), ('$const0.2 = const(int, 1)  :: Literal[int](1)', ''), ('$0.3 = x + $const0.2  :: int64', ''), ('del x', ''), ('del $const0.2', ''), ('$0.4 = cast(value=$0.3)  :: int64', ''), ('del $0.3', ''), ('return $0.4', '')]}, 'ir_indent': {5: [0], 6: [], 7: [2, 2, 2, 2, 2, 2, 2, 2]}})])

 Second signature
OrderedDict([(('_examples/typeannotation-global.py:6', '(float64,) -> float64'), {'num_lifted_loops': 0, 'filename': '_examples/typeannotation-global.py', 'funcname': 'foo', 'python_lines': [(5, '@njit'), (6, 'def foo(x):'), (7, 'return x + 1')], 'python_indent': {5: 0, 6: 0, 7: 4}, 'python_tags': {5: '', 6: '', 7: ''}, 'ir_lines': {5: [('label 0', '')], 6: [], 7: [('x = arg(0, name=x)  :: float64', ''), ('$const0.2 = const(int, 1)  :: Literal[int](1)', ''), ('$0.3 = x + $const0.2  :: float64', ''), ('del x', ''), ('del $const0.2', ''), ('$0.4 = cast(value=$0.3)  :: float64', ''), ('del $0.3', ''), ('return $0.4', '')]}, 'ir_indent': {5: [0], 6: [], 7: [2, 2, 2, 2, 2, 2, 2, 2]}})])
```
</details>


Happy to modify things here if I'm missing something

stuartarchibald commented:
Thanks for proposing these changes, I have a recollection of noticing what you describe, and a somewhat vague recollection of there being a bug report/comment about it but unfortunately can't find it. IIRC `get_annotation_info` has the wrong behaviour but will need to take a further look to see if there's any side effects of making the proposed fix, assuming all is well, thanks for chasing this down and fixing it!

stuartarchibald commented:
I've now queued this for review. Note to reviewer(s)/self, probably worth checking the `numba` binary manually as well use in `jupyter-notebook`/`ipython REPL`. Thanks.

jrbourbeau commented:
Awesome, thanks for the update @stuartarchibald

stuartarchibald commented on numba/pretty_annotate.py:
```diff
@@ -246,13 +246,12 @@ class Annotate:
     Function annotations persist across compilation for newly encountered
     type signatures and as a result annotations are shown for all signatures.
```
I think this docstring might need an update now?

jrbourbeau replied:
Just updated to state that annotations for all signatures are shown by default, and included an example of using the `signature` keyword

stuartarchibald commented on numba/tests/test_dispatcher.py:
```diff
@@ -859,6 +861,31 @@ def foo(a, b):
+        with captured_stdout() as second:
+            foo.inspect_types(signature=foo.signatures[1])
+
+        assert total.getvalue() == first.getvalue() + second.getvalue()
```
Numba uses standard `unittest` behaviours for assertions, e.g. `self.assertEqual()` or `self.assertTrue()` would be appropriate in this case.

jrbourbeau replied:
Thanks for pointing that out, I've updated accordingly

stuartarchibald commented on numba/tests/test_dispatcher.py:
```diff
@@ -859,6 +861,31 @@ def foo(a, b):
+            foo.inspect_types()
+        assert expected in out.getvalue()
+
+    def test_inspect_types_signature(self):
```
```suggestion
    def test_inspect_types_with_signature(self):
```

stuartarchibald commented on numba/tests/test_dispatcher.py:
```diff
@@ -887,6 +914,19 @@ def foo(a, b):
+        expected = dict(chain.from_iterable(foo.get_annotation_info(i).items()
+                                            for i in foo.signatures))
+        result = foo.get_annotation_info()
+        assert expected == result
```
Same comment as above RE assertions.

jrbourbeau commented:
Thanks for the review @stuartarchibald! I pushed some commits that should address your comments

stuartarchibald commented:
@jrbourbeau thanks for the fixes. This can be marked for merge once CI passes.

stuartarchibald commented:
(and congratulations on your first code contribution to Numba!)
